src/backend/gporca/libnaucrates/src/statistics/CBucket.cpp

```diff
@@ -415,13 +415,8 @@ CBucket::MakeBucketSingleton(CMemoryPool *mp, CPoint *point_singleton) const
 	// if the bucket is not already singleton, scale accordingly
 	if (!this->IsSingleton())
 	{
-		CDouble ratio =
-			CDouble(1.0) / m_bucket_upper_bound->Width(m_bucket_lower_bound,
-													   m_is_lower_closed,
-													   m_is_upper_closed);
-		frequency_new =
-			std::min(DOUBLE(1.0), (this->m_frequency * ratio).Get());
-		;
+		// Calculate new frequency based on NDV ratio
+		frequency_new = std::min(DOUBLE(1.0), (m_frequency / m_distinct).Get());
```

Comment:
Just wondering whether we could ever have an empty bucket with a frequency and distinct of 0? The other thing I'm wondering is whether we should worry about the case where 0 < m_distinct < 1.

```suggestion
		// scale NDV down to 1 (or take the entire NDV if it's less than 1),
		// then scale the frequency by the same ratio
		CDouble ratio = std::min(1.0, m_distinct);
		distinct_new = m_distinct / ratio;
		frequency_new = m_frequency / ratio;
```

Reply:
Yes, that makes sense to handle empty bucket.  I'm not as sure about the second case though.  If 0 < m_distinct < 1, then is it still singleton bucket?

Reply:
> If 0 < m_distinct < 1, then is it still singleton bucket?

Yes, but I would suggest that instead of an NDV of 1, we use the NDV that's less than 1 so that the result doesn't have a higher NDV and frequency than the input argument. That assumes that we can have an NDV of < 1 in a bucket (both in a non-singleton and a singleton bucket).

Reply:
Ah, got it. Thanks for that clarity